Utils/Services/ImagePickerService.swift

```diff
@@ -0,0 +1,141 @@
+    
+    /**
+     Tells if the device has a camera.
+     */
+    var cameraIsAvailable: Bool { get }
+}
+
+@objc
+public final class ImagePickerService: NSObject, ImagePickerServiceType {
+    
+    public let imageSignal: Signal<UIImage, ImagePickerServiceError>
+    fileprivate let _imageObserver: Signal<UIImage, ImagePickerServiceError>.Observer
+    
+    fileprivate weak var _viewController: UIViewController?
+    
+    init(viewController: UIViewController) {
+        _viewController = viewController
+        (imageSignal, _imageObserver) = Signal<UIImage, ImagePickerServiceError>.pipe()
+    }
+    
+    public func presentCameraTypeImagePickerController(_ permissionNotGranted: @escaping (Void) -> Void) {
+        if cameraIsAvailable {
+            hasCameraPermission().startWithValues { [unowned self] in
+                $0 ? self.presentImagePickerController(.camera) : permissionNotGranted()
+            }
+        } else {
+            _imageObserver.send(error: ImagePickerServiceError.sourceTypeNotAvailable)
+        }
+    }
+    
+    public func presentGalleryTypeImagePickerController(_ permissionNotGranted: @escaping (Void) -> Void) {
+        hasPhotosPermission().startWithValues { [unowned self] in
+            $0 ? self.presentImagePickerController(.photoLibrary) : permissionNotGranted()
+        }
+    }
+    
+    public var cameraIsAvailable: Bool {
+        return UIImagePickerController.isSourceTypeAvailable(.camera)
+    }
+    
+    deinit {
+        _imageObserver.sendCompleted()
+    }
+}
+
+fileprivate extension ImagePickerService {
+    
+    fileprivate func hasCameraPermission() -> SignalProducer<Bool, NoError> {
+        return SignalProducer { observable, _ in
+            switch AVCaptureDevice.authorizationStatus(forMediaType: AVMediaTypeVideo) {
+            case .notDetermined:
+                AVCaptureDevice.requestAccess(forMediaType: AVMediaTypeVideo) {
+                    observable.send(value: $0)
+                }
+            case .authorized: observable.send(value: true)
+            case .denied, .restricted: observable.send(value: false)
+            }
+        }
+    }
+    
+    fileprivate func hasPhotosPermission() -> SignalProducer<Bool, NoError> {
+        return SignalProducer { observable, _ in
+            switch PHPhotoLibrary.authorizationStatus() {
+            case .notDetermined:
+                PHPhotoLibrary.requestAuthorization {
+                    observable.send(value: $0 == .authorized)
+                }
+            case .authorized: observable.send(value: true)
+            case .denied, .restricted: observable.send(value: false)
+            }
+        }
+    }
+    
+}
+
+extension ImagePickerService: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
+    
+    public func imagePickerController(_ picker: UIImagePickerController,
+                               didFinishPickingImage image: UIImage, editingInfo: [String : AnyObject]?) {
```

Comment:
I would indent this better, although I'm not sure what would be better.
I guess have 3 lines, and align `didFinishPickingImage` and `editingInfo` with `_ picker`.